tgui/packages/tgui/interfaces/AntagInfoHeretic.tsx

```diff
@@ -219,63 +232,75 @@ const InformationSection = (props) => {
   );
 };
 
-const ResearchedKnowledge = (props) => {
-  const { data } = useBackend<KnowledgeInfo>();
-  const { learnedKnowledge } = data;
-
-  return (
-    <Stack.Item grow>
-      <Section title="Researched Knowledge" fill scrollable>
-        <Stack vertical>
-          {(!learnedKnowledge.length && 'None!') ||
-            learnedKnowledge.map((learned) => (
-              <Stack.Item key={learned.name}>
-                <Button
-                  width="100%"
-                  color={learned.color}
-                  content={`${learned.hereticPath} - ${learned.name}`}
-                  tooltip={learned.desc}
-                />
-              </Stack.Item>
-            ))}
-        </Stack>
-      </Section>
-    </Stack.Item>
-  );
-};
-
-const KnowledgeShop = (props) => {
+const KnowledgeTree = (props) => {
   const { data, act } = useBackend<KnowledgeInfo>();
-  const { learnableKnowledge } = data;
+  const { knowledge_tiers } = data;
 
   return (
-    <Stack.Item grow>
-      <Section title="Potential Knowledge" fill scrollable>
-        {(!learnableKnowledge.length && 'None!') ||
-          learnableKnowledge.map((toLearn) => (
-            <Stack.Item key={toLearn.name} mb={1}>
-              <Button
-                width="100%"
-                color={toLearn.color}
-                disabled={toLearn.disabled}
-                content={`${toLearn.hereticPath} - ${
-                  toLearn.cost > 0
-                    ? `${toLearn.name}: ${toLearn.cost}
-                  point${toLearn.cost !== 1 ? 's' : ''}`
-                    : toLearn.name
-                }`}
-                tooltip={toLearn.desc}
-                onClick={() => act('research', { path: toLearn.path })}
-              />
-              {!!toLearn.gainFlavor && (
-                <BlockQuote>
-                  <i>{toLearn.gainFlavor}</i>
-                </BlockQuote>
-              )}
+    <Section title="Research Tree" fill scrollable>
+      <Stack vertical>
+        {(!knowledge_tiers.length && 'None!') ||
+          knowledge_tiers.map((tier, i) => (
+            <Stack.Item key={i}>
+              <Flex
+                direction="row"
+                justify="center"
+                align="center"
+                backgroundColor="#303030"
+                wrap="wrap"
+              >
+                {
+                  tier.nodes.map((node) => (
+                    <Flex.Item key={node.name}>
+                      <Button key={node.name}
+                        color={node.color}
+                        disabled={node.disabled}
+                        tooltip={
+                          `${node.name}:
+                          ${node.desc}`
+                        }
+                        onClick={node.finished ? (() => null) : () => act('research', { path: node.path })}
```

Comment:
Make this a function pls

Reply:
You mean move the checks into the .dm file?

Reply:
nono, it's just that i like to keep the complexity out of the jsx block. i meant to define this as a `function onClickHandler()` and pass it into `onClick={onClickHandler}`

Reply:
A question, sorry if this is obvious, I don't have much experience with jsx - I'm not sure how to pass the `node` to the `onClickHandler`, seems like `onClick` always calls it with no parameters. Can I somehow define the function within the scope where `node` exists?

Reply:
Figured it out, but I've been also told keeping this line the original way should be better, so I'll keep it that way probably